src/raftstore/store/worker/apply.rs

```diff
@@ -195,48 +194,126 @@ impl ApplyCallback {
 
 struct ApplyContext<'a> {
     host: &'a CoprocessorHost,
-    wb: WriteBatch,
+    wb: Option<WriteBatch>,
     cbs: MustConsumeVec<ApplyCallback>,
+    apply_res: Vec<ApplyRes>,
     wb_last_bytes: u64,
     wb_last_keys: u64,
-    sync_log: bool,
+    committed_count: usize,
+    // `enable_sync_log` is a global switch indicate that wal can be synchronized
+    // when data is written to kv engine.
+    enable_sync_log: bool,
+    // `sync_log_hint` indicates whether synchronize wal is prefered.
+    sync_log_hint: bool,
     exec_ctx: Option<ExecContext>,
     use_delete_range: bool,
 }
 
 impl<'a> ApplyContext<'a> {
-    fn new(host: &CoprocessorHost, use_delete_range: bool) -> ApplyContext {
+    fn new(host: &CoprocessorHost) -> ApplyContext {
         ApplyContext {
             host: host,
-            wb: WriteBatch::with_capacity(DEFAULT_APPLY_WB_SIZE),
+            wb: None,
             cbs: MustConsumeVec::new("callback of apply context"),
+            apply_res: vec![],
             wb_last_bytes: 0,
             wb_last_keys: 0,
-            sync_log: false,
+            committed_count: 0,
+            enable_sync_log: false,
+            sync_log_hint: false,
             exec_ctx: None,
-            use_delete_range: use_delete_range,
+            use_delete_range: false,
         }
     }
 
+    fn enable_sync_log(mut self, eanbled: bool) -> ApplyContext<'a> {
+        self.enable_sync_log = eanbled;
+        self
+    }
+
+    fn apply_res_capacity(mut self, cap: usize) -> ApplyContext<'a> {
+        self.apply_res = Vec::with_capacity(cap);
+        self
+    }
+
+    fn use_delete_range(mut self, use_delete_range: bool) -> ApplyContext<'a> {
+        self.use_delete_range = use_delete_range;
+        self
+    }
+
+    /// Prepare for applying entries for `delegate`.
+    ///
+    /// A general apply progress for a delegate is:
+    /// `prepare_for` -> `commit` [-> `commit` ...] -> `finish_for`.
+    /// After all delegates are handled, `write_to_db` method should be called.
     fn prepare_for(&mut self, delegate: &ApplyDelegate) {
+        if self.wb.is_none() {
+            self.wb = Some(WriteBatch::with_capacity(DEFAULT_APPLY_WB_SIZE));
+        }
         self.cbs.push(ApplyCallback::new(delegate.region.clone()));
     }
 
-    pub fn mark_last_bytes_and_keys(&mut self) {
-        self.wb_last_bytes = self.wb.data_size() as u64;
-        self.wb_last_keys = self.wb.count() as u64;
+    /// Commit all changes have done for delegate. `persistent` indicates whether
+    /// write the changes into rocksdb.
+    ///
+    /// This call is valid only when it's between a `prepare_for` and `finish_for`.
+    fn commit(&mut self, delegate: &mut ApplyDelegate, persistent: bool) {
+        if persistent {
```

Comment:
I prefer to merge this if and the below if, although the apply state will not be included in statistics.

Reply:
That's the exact reason why they should not be merged.